# Homework(1)_ML_Zoomcamp

# 1-Pandas Version

In [1]:
import pandas as pd
pd.__version__

'2.3.1'

# Getting the data

In [2]:
df =pd.read_csv('/workspaces/ML-ZoomCamp/car_fuel_efficiency.csv')
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

# 2-Records count

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


Records count : 9704

# 3- Number of columns having null values

In [4]:
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

There are four columns having missing values

# 4- Fuel Types

In [5]:
df['fuel_type'].value_counts()

fuel_type
Gasoline    4898
Diesel      4806
Name: count, dtype: int64

There are two fuel types in the dataset : Gasoline and Diessel

# 5-Maximum fuel efficiency of cars from Asia

In [6]:
df['origin'].value_counts()

origin
Europe    3254
Asia      3247
USA       3203
Name: count, dtype: int64

In [7]:
max_fuel_asia = df.loc[df['origin'] == 'Asia', 'fuel_efficiency_mpg'].max()
print(max_fuel_asia)


23.759122836520497


# 6- Median value of horsepower

In [8]:
df['horsepower'].describe()

count    8996.000000
mean      149.657292
std        29.879555
min        37.000000
25%       130.000000
50%       149.000000
75%       170.000000
max       271.000000
Name: horsepower, dtype: float64

median of horsepower = 149

In [9]:
frequent_horse_power = df['horsepower'].mode()[0]
print(frequent_horse_power)

152.0


In [10]:
complete_horse_power = df['horsepower'].fillna(frequent_horse_power, inplace=False)

In [11]:
complete_horse_power.describe()

count    9704.000000
mean      149.828215
std        28.775248
min        37.000000
25%       131.000000
50%       152.000000
75%       168.000000
max       271.000000
Name: horsepower, dtype: float64

new median of horsepower = 152  It increased after filling the missing values

# 7-Sum of weights calculation

In [12]:
car_asia_df = df[df['origin'] == 'Asia']
car_asia_df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0         NaN     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [13]:
# select only columns vehicle_weight and model_year
car_asia_df[['vehicle_weight', 'model_year']][0:7]

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [14]:
X = car_asia_df[['vehicle_weight', 'model_year']][0:7].values
print(X)

[[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]


In [15]:
X_transpose = X.T
print(X_transpose)

[[2714.21930965 2783.86897424 3582.68736772 2231.8081416  2659.43145076
  2844.22753389 3761.99403819]
 [2016.         2010.         2007.         2011.         2016.
  2014.         2019.        ]]


In [16]:
XTX = X_transpose.dot(X)
print(XTX)

[[62248334.33150762 41431216.5073268 ]
 [41431216.5073268  28373339.        ]]


In [17]:
import numpy as np
XTX_inv = np.linalg.inv(XTX)
print(XTX_inv)


[[ 5.71497081e-07 -8.34509443e-07]
 [-8.34509443e-07  1.25380877e-06]]


In [18]:
y = [1100, 1300, 800, 900, 1000, 1100, 1200]

In [19]:
w= XTX_inv.dot(X_transpose).dot(y)
print(w)

[0.01386421 0.5049067 ]


In [20]:
w.sum()

np.float64(0.5187709081074016)

sum of Weights = 0.51